### PR TITLE
Weave Net Monitoring setup using kube-prometheus

### PR DESCRIPTION
**Pull request by alok87**
Fixes https://github.com/coreos/kube-prometheus/issues/424

## Contents
Using kube-prometheus and kubectl you will be able install the following for monitoring weave-net in your cluster:

- [x]  [Service for Weave Net](https://gist.github.com/alok87/379c6234b582f555c141f6fddea9fbce) The service which the [service monitor](https://coreos.com/operators/prometheus/docs/latest/user-guides/cluster-monitoring.html) scraps.
- [x] [ServiceMonitor for Weave Net](https://gist.github.com/alok87/e46a7f9a79ef6d1da6964a035be2cfb9) Service monitor to scraps the weavenet metrics and bring it to Prometheus.
- [x] [Prometheus Alerts for Weave Net](https://stackoverflow.com/a/60447864) This will setup all the important weave net metrics you should be alerted on.
- [x] [Grafana Dashboard for Weave Net](https://grafana.com/grafana/dashboards/11789) This will setup the per CNI pod level monitoring for weave net.
- [x] [Grafana Dashboard for Weave Net(Cluster)](https://grafana.com/grafana/dashboards/11804) This will setup the cluster level monitoring for weave net.

cc @bboreham

### CODE REVIEW

**Comment by alok87** [comment]
@brancz how do i add dashboards for grafana? Can you suggest some doc or referrence PR. The dasboards are open source and are hosted in grafana.com

**Comment by alok87** [comment]
Not sure why the CI is failing, does not look related to this PR.

**Comment by alok87** [comment]
Please suggest how to fix the CI 
```
go/bin/jb install
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5ebf57]
```

**Comment by brancz** [comment]
Generally looking good to me not sure what the CI failure is though.

**Comment by alok87** [comment]
@brancz got passed. The issue got fixed after this https://github.com/coreos/kube-prometheus/pull/425/commits/901e7a9a148cfe202843e6b8c6fcdf45be9086cd

**Comment by alok87** [comment]
Rebased with the master.
Added an example to customise alerts.
Please take it forward.

File: docs/weave-net-support.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+Using kube-prometheus and kubectl you will be able install the following for monitoring weave-net in your cluster:
+
+1. [Service for WeaveNet](https://gist.github.com/alok87/379c6234b582f555c141f6fddea9fbce) The service which the [service monitor](https://coreos.com/operators/prometheus/docs/latest/user-guides/cluster-monitoring.html) scraps.
+2. [ServiceMonitor for WeaveNet](https://gist.github.com/alok87/e46a7f9a79ef6d1da6964a035be2cfb9) Service monitor to scraps the weavenet metrics and bring it to Prometheus.
+3. [Prometheus Alerts for WeaveNet](https://stackoverflow.com/a/60447864) This will setup all the important weave net metrics you should be alerted on.
+4. [Grafana Dashboard for WeaveNet](https://grafana.com/grafana/dashboards/11789) This will setup the per CNI pod level monitoring for weave net.
+5. [Grafana Dashboard for WeaveNet(Cluster)](https://grafana.com/grafana/dashboards/11789) This will setup the cluster level monitoring for weave net.
+
+## Instructions
+- You can monitor weave-net CNI using kube-prometheus with:
```
Comment (brancz) [inline docs/weave-net-support.md]: Can you clarify that the alert medication is not required? People tend to just copy these things :)

Comment (alok87) [reply docs/weave-net-support.md]: Few alerts are environment specific. I will bring the ones which may require change in the  example and also add a note saying we may need to change it. Thanks 👀

Comment (alok87) [reply docs/weave-net-support.md]: Done

File: docs/weave-net-support.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Setup WeaveNet CNI monitoring using kube-prometheus
+[WeaveNet](https://kubernetes.io/docs/concepts/cluster-administration/networking/#weave-net-from-weaveworks) is a resilient and simple to use CNI for Kubernetes. A well monitored and observed CNI helps in troubleshooting Kubernetes networking problems. [WeaveNet](https://www.weave.works/docs/net/latest/concepts/how-it-works/) emits [prometheus metrics](https://www.weave.works/docs/net/latest/tasks/manage/metrics/) for monitoring WeaveNet. There are many ways to install WeaveNet in your cluster. One of them is using [kops](https://github.com/kubernetes/kops/blob/master/docs/networking.md).
+
+Following this document, you can setup weave net CNI monitoring for your cluster using kube-prometheus.
```
Comment (squat) [inline docs/weave-net-support.md]: Is it `WeaveNet` or `weave net`?

File: docs/weave-net-support.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# Setup WeaveNet CNI monitoring using kube-prometheus
+[WeaveNet](https://kubernetes.io/docs/concepts/cluster-administration/networking/#weave-net-from-weaveworks) is a resilient and simple to use CNI for Kubernetes. A well monitored and observed CNI helps in troubleshooting Kubernetes networking problems. [WeaveNet](https://www.weave.works/docs/net/latest/concepts/how-it-works/) emits [prometheus metrics](https://www.weave.works/docs/net/latest/tasks/manage/metrics/) for monitoring WeaveNet. There are many ways to install WeaveNet in your cluster. One of them is using [kops](https://github.com/kubernetes/kops/blob/master/docs/networking.md).
```
Comment (squat) [inline docs/weave-net-support.md]: Nit, capitalize "Prometheus" in `emits prometheus metrics`

Comment (squat) [inline docs/weave-net-support.md]: > simple to use CNI

Do you mean "CNI plugin" or maybe "CNI provider"? CNI is an interface, not a concrete thing

File: docs/weave-net-support.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Following this document, you can setup weave net CNI monitoring for your cluster using kube-prometheus.
+
+## Contents
+Using kube-prometheus and kubectl you will be able install the following for monitoring weave-net in your cluster:
```
Comment (squat) [inline docs/weave-net-support.md]: This introduces another variation: "weave-net"

File: docs/weave-net-support.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+1. [Service for WeaveNet](https://gist.github.com/alok87/379c6234b582f555c141f6fddea9fbce) The service which the [service monitor](https://coreos.com/operators/prometheus/docs/latest/user-guides/cluster-monitoring.html) scraps.
```
Comment (squat) [inline docs/weave-net-support.md]: s/scraps/scrapes/

File: docs/weave-net-support.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+2. [ServiceMonitor for WeaveNet](https://gist.github.com/alok87/e46a7f9a79ef6d1da6964a035be2cfb9) Service monitor to scraps the weavenet metrics and bring it to Prometheus.
```
Comment (squat) [inline docs/weave-net-support.md]: s/scraps/scrapes/

File: docs/weave-net-support.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+3. [Prometheus Alerts for WeaveNet](https://stackoverflow.com/a/60447864) This will setup all the important weave net metrics you should be alerted on.
```
Comment (squat) [inline docs/weave-net-support.md]: s/weave net/WeaveNet/

File: docs/weave-net-support.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+4. [Grafana Dashboard for WeaveNet](https://grafana.com/grafana/dashboards/11789) This will setup the per CNI pod level monitoring for weave net.
```
Comment (squat) [inline docs/weave-net-support.md]: s/weave net/WeaveNet/

File: docs/weave-net-support.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+5. [Grafana Dashboard for WeaveNet(Cluster)](https://grafana.com/grafana/dashboards/11789) This will setup the cluster level monitoring for weave net.
```
Comment (squat) [inline docs/weave-net-support.md]: s/weave net/WeaveNet/

File: docs/weave-net-support.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+## Instructions
+- You can monitor weave-net CNI using an example like below. **Please note that some alert configurations are environment specific and may require modifications of alert thresholds**. For example: The FastDP flows have never gone below 1500 for us. But if this value is say 2000 for you then you can use an example like below to update the alert. The alerts which may require threshold modifications are `WeaveNetFastDPFlowsLow` and `WeaveNetIPAMUnreachable`.
```
Comment (squat) [inline docs/weave-net-support.md]: s/weave-net/WeaveNet/

**Comment by alok87** [comment]
@squat in the code i see it is always referred as `weave-net` and in the documentation as `WeaveNet`. I am following this convention to make the changes.

File: docs/weave-net-support.md (hunk 3 above)
Comment (squat) [reply docs/weave-net-support.md]: This is still outstanding, it doesn't really mean anything to be `a CNI`

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+          {
+            alert: 'WeaveNetIPAMUnreachable',
+            expr: 'weave_ipam_unreachable_percentage > 25',
+            'for': '10m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'WeaveNetIPAM unreachability percentage is above threshold. Go to the below prometheus link for details.',
+              description: 'Actionable: Find why the unreachability threshold have increased from threshold and fix it. WeaveNet is responsible to keep it under control. Weave rm peer deployment can help clean things.',
```
[the remark this answers is not in the document]
Comment (squat) [reply jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: `have` does not match with the subject. Should be `has` or it should be a plural subject

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetIPAMPendingClaims',
+            expr: 'sum(weave_ipam_pending_claims) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'WeaveNet IPAM has pending claims. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: Same comment about `WeaveNetIPAM`

Comment (squat) [reply jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetFastDPFlowsLow',
+            expr: 'sum(weave_flows) < 15000',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'WeaveNet total FastDP flows is below threshold. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/WeaveNet/Weave Net/

Comment (squat) [reply jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+              description: 'Actionable: Find the reason for fast dp flows dropping below the threshold.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: Is it `FastDP` or `fast dp`?

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetFastDPFlowsOff',
+            expr: 'sum(weave_flows == bool 0) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'WeaveNet FastDP flows is not happening in some or all nodes. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/WeaveNet/Weave Net/

Comment (squat) [reply jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+              description: 'Actionable: Find the reason for fast dp being off.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: should be `fast dp flows`

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetHighConnectionTerminationRate',
+            expr: 'rate(weave_connection_terminations_total[5m]) > 0.1',
+            'for': '5m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'A lot of connections are getting terminated. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetConnectionsConnecting',
+            expr: 'sum(weave_connections{state="connecting"}) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'A lot of connections are in connecting state. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetConnectionsRetying',
+            expr: 'sum(weave_connections{state="retrying"}) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'A lot of connections are in retrying state. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetConnectionsPending',
+            expr: 'sum(weave_connections{state="pending"}) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'A lot of connections are in pending state. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+          },
+          {
+            alert: 'WeaveNetConnectionsFailed',
+            expr: 'sum(weave_connections{state="failed"}) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'A lot of connections are in failed state. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+        rules: [
+          {
+            alert: 'WeaveNetIPAMSplitBrain',
+            expr: 'max(weave_ipam_unreachable_percentage) - min(weave_ipam_unreachable_percentage) > 0',
+            'for': '3m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'WeaveNetIPAM has a split brain. Go to the below prometheus link for details.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: s/prometheus/Prometheus/

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+              description: 'Actionable: Every node should see same unreachability percentage. Please check and fix why it is not so.',
```
Comment (squat) [inline jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: "should see *the* same ..."

**Comment by alok87** [comment]
Not changing the alert names because it is in accordance with the existing alert conventions.  https://github.com/coreos/kube-prometheus/pull/425#discussion_r387719890
Screenshot 2020-03-04 at 8 42 12 PM

File: jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet (hunk 11 above)
[the remark this answers is not in the document]
Comment (alok87) [reply jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: https://github.com/coreos/kube-prometheus/pull/425#issuecomment-594588992

[the remark this answers is not in the document]
Comment (squat) [reply jsonnet/kube-prometheus/kube-prometheus-weave-net.libsonnet]: yes no need to change the names, just standardize the descriptions and summaries

File: docs/weave-net-support.md [hunk 24]
```diff
@@ -0,0 +1,67 @@
+# Setup Weave Net CNI monitoring using kube-prometheus
+[Weave Net](https://kubernetes.io/docs/concepts/cluster-administration/networking/#weave-net-from-weaveworks) is a resilient and simple to use CNI for Kubernetes. A well monitored and observed CNI helps in troubleshooting Kubernetes networking problems. [Weave Net](https://www.weave.works/docs/net/latest/concepts/how-it-works/) emits [prometheus metrics](https://www.weave.works/docs/net/latest/tasks/manage/metrics/) for monitoring Weave Net. There are many ways to install Weave Net in your cluster. One of them is using [kops](https://github.com/kubernetes/kops/blob/master/docs/networking.md).
```
Comment (squat) [inline docs/weave-net-support.md]: Ping again about this. It's awkward to call Weave Net a `CNI`. As I mentioned before, CNI is an interface, it is not a concrete thing. Consider something like "CNI provider"

Comment (alok87) [reply docs/weave-net-support.md]: pushing

Comment (alok87) [reply docs/weave-net-support.md]: done

**Pull request merged**